doc/error-handling.md

```diff
@@ -0,0 +1,324 @@
+```
+A line break is inserted implicitly at the end (after "log message continued" in the example).
+
+## Error Handler
+The most general use case is the following
+```
+if(noError()) {
+    //handle the regular case
+} else {
+    auto callback = []() { //some error handling callback};
+    errorHandler(Error::kSOME_ERROR_CODE, callback, ErrorLevel::SEVERE);
+}
+```
+
+If no callback or error level are specified, the error level is assumed to be FATAL by default.
+```
+errorHandler(Error::kSOME_ERROR_CODE)
+```
+
+If no callback but an error level is desired, a nullptr has to be provided for the callback.
+```
+errorHandler(Error::kSOME_ERROR_CODE, nullptr, ErrorLevel::MODERATE);
+```
+
+## Expects and Ensures
+
+Assume func is part of an inner API and not supposed to be called with a nullptr. We may have used a reference here, this is just for illustration.
+In addition the value pointed to is assumed to be in the range (-1024, 1024). While we could check this everytime, this can be avoided if we specify that the caller is responsible to ensure that these conditions hold.
+
+```
+int myAlgorithm(int* ptr) {
+    Expects(ptr!=nullptr);
+    //observe the order, we only dereference after the nullptr check
+    Expects(*ptr > -1024 && *ptr < 1024);
+
+    int intermediate = timesTwo(*ptr);
+    //this may not be necessary here to ensure that the next function call is valid,
+    //but it states our expectations clearly
+    Ensures(intermediate % 2 == 0);
+
+    int result = abs(intermediate);
+
+    Ensures(result % 2 == 0);
+    Ensures(result >= 0);
+    Ensures(result < 2048);
+
+    return result;
+}
+```
+Note that in the case of nullptr checks it is also an option to use references in arguments (or **not_null** if it is supposed to be stored since references are not copyable). It should be considered that not_null incurs a runtime cost, which may be undesirable.
+When Expects and Ensures are implemented to leave no trace in Release Mode, we do not incur a runtime cost using them. For this reason it is advised to use them to document and verify assumptions where approriate.
+
+## cxx::expected
+This example checks the arguments and if they are valid proceeds to compute a result and returns it.
+Otherwise it creates an Error object from an errorCode and returns it.
+ 
+```
+std::expected<SomeType, Error>::func(Arg arg) {
+    int errorCode = checkArg(arg);
+    if(isNoError(errorCode)) {
+        SomeType result = computeResult(arg);
+        // optionally do something with result
+        return result;
+    }
+    return Error(errorCode);
+}
+```
+The caller is responsible for handling (or propagating) the error.
+
+```
+auto result = func(arg);
+if(result.has_error()) {
+    auto error = result.error();
+    //handle or propagate the error
+} else {
+    auto value = result.value();
+    //proceed by using the value
+}
+```
+
+Alternatively a functional approach can be used.
+
+```
+auto successFunc = [](SomeType& result) { 
```

Comment:
this is obsolete, nowadays you get a reference to the actual value in `and_then` or error value in `or_else`

Reply:
Please clarify, the lambda takes a reference to the value (SomeType in this case) as you wrote. Similarly the errorFunc takes Error&.

Reply:
Note that by using some more functional conventions an even better chaining of functions on multiple results could be achieved. Nevertheless, this is a bit awkward since the language does not natively support those concepts. I can demonstrate my best effort at this sometime. :-)